plat: Migrate pl011 to drivers/uktty

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [ ] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [N/A]
 - Platform(s): [N/A]
 - Application(s): [N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Migrate the PL011 driver to `drivers/uktty`

michpappas commented:
Notice that the driver is introduced as a platform library, ie:
```
$(eval $(call addplatlib_s,kvm,libuktty_pl011,$(CONFIG_LIBUKTTY_PL011)))
```
This is due to limitations of #1023. **Do not merge until resolved.**

michpappas commented:
Changes in this update:
- Rebase to latest state of #1023
- Replace subsystem's `Config.uk` `menuconfig` with `menu`
- Register library using `addlib_s`
- Add `exportsyms.uk`
- Add `ASINCLUDES` and `CXXINCLUDES`

michpappas committed:
drivers: Introduce uktty for console devices

Add skeleton for the `uktty` subsystem. This subsystem contains
drivers for console devices such as UART and VGA.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/drivers: Migrate pl011 to drivers/uktty

Migrate the pl011 driver to the newly introduced uktty subsystem.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/pl011: Replace stdint types with kernel types

The `pl011` does not depend on the standand library. Replace stdint
types with kernel types.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/pl011: Use UK_BIT in bit shift macros

Replace bit shifts with UK_BIT in macros to comply with checkpath.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/pl011: Whitespace fixes for checkpatch

Various whitespace fixes to comply with checkpatch.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/uktty/pl011: Use shorthand syntax on NULL check

Use shorhand syntax for checks against NULL to comply with checkpatch.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/tty/pl011: Set the initialized flag when configuring via fdt

Fixes an issue when early console is not enabled where no output
would be sent to the console due to the driver ignoring calls when
the initialized flag is not set.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

razvanvirtan reviewed (approved):
Looks good to me, just a few typos in the commit messages:

- https://github.com/unikraft/unikraft/pull/1025/commits/f2ddbdb514cddab21aaad404595f5dbc4673467f `console`
- https://github.com/unikraft/unikraft/pull/1025/commits/86e71403de1f045f4017d28d05335f992a3f704c `chackpatch`

Approved-by: Razvan Virtan <virtanrazvan@gmail.com>

michpappas commented:
@razvanvirtan thanks a lot for reviewing. Both typos are fixed.

mogasergiu commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -128,7 +128,7 @@ void pl011_console_init(const void *dtb)
 
 	regs = fdt_getprop(dtb, offset, "reg", &len);
-	if (regs == NULL || (len < (int)sizeof(fdt32_t) * (naddr + nsize)))
+	if (!regs || (len < (int)sizeof(fdt32_t) * (naddr + nsize)))
```
Let's use `unlikely` on all `UK_CRASH` cases?

michpappas replied:
Good point, I'll do that. Thanks 👍🏼

mogasergiu commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -90,7 +90,7 @@ static void init_pl011(__u64 bas)
 	/* Mask all interrupts */
 	PL011_REG_WRITE(REG_UARTIMSC_OFFSET,
-		PL011_REG_READ(REG_UARTIMSC_OFFSET) & 0xf800);
+			PL011_REG_READ(REG_UARTIMSC_OFFSET) & 0xf800);
```
Could we maybe have `drivers/uktty/pl011: Whitespace fixes for checkpatch` squashed into the migration commit? :thinking:

michpappas replied:
It's also one possible way of many. I considered both that option as well as the possibility to do all checkpatch changes in one commit (or even in the same one as the migration), but in the end I concluded it would be cleaner to do this in discrete steps. If you have a strong opinion / reasoning about it I don't mind to change it though.

mogasergiu replied:
> If you have a strong opinion / reasoning about it I don't mind to change it though.

Nah, my only concern is usually when it comes to bisecting and looking at git history to not be unnecessarily contaminated.

> I concluded it would be cleaner to do this in discrete steps

Ok I see. Sure, it's just one commit after all. Let's keep it like this if you say you considered it already and believe it to be cleaner: migration and checkpatch fixes. :+1:

mogasergiu reviewed (commented):
Hi @michpappas! Thanks for not letting this PR rot :pray: ! Had a quick look over the changes. I will have another run tomorrow on this and the others from your side I am assigned as reviewer to.

michpappas committed:
drivers/uktty/pl011: Use `unlikely` in error conditions

Use the `unlikely` macro in error conditions to hint the compiler
to optimize the for the probable path.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented on drivers/uktty/pl011/pl011.c:
```diff
@@ -70,9 +69,9 @@
 static uint8_t pl011_uart_initialized = 1;
-static uint64_t pl011_uart_bas = CONFIG_EARLY_PRINT_PL011_UART_ADDR;
+static uint64_t pl011_uart_bas = CONFIG_LIBUKTTY_PL011_EARLY_CONSOLE_BASE;
 #else
```
An `/* !CONFIG_LIBUKTTY_PL011_EARLY_CONSOLE_BASE */` here?

mogasergiu reviewed (approved):
Just a tiny comment. Other than that, LGTM!

Reviewed-by: Sergiu Moga <sergiu.moga@unikraft.io>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
